[WIP]: make hw/ directory configurable

yangzhon opened the pull request:
@sameo @rbradford @mcastelino 

This PR make all files in hw/ directory configurable and make new x86_64_virt-softmmu.mak for virt-i386 platform. With this new platform mak file, we can make specific binary for virt platform.

How to build binary with this new mak file
./configure --target-list=x86_64_virt-softmmu

Currently, there are some compile error because i removed some CONFIG_XXX in the third patch, please help check those patches, especially for the new x86_64_virt-softmmu.mak file, which will be less CONFIG_XXX enabled.

Let me list the current CONFIG_XXX here:
CONFIG_PCI=y
CONFIG_VIRTIO_PCI=y
CONFIG_VIRTIO=y
CONFIG_ACPI=y
CONFIG_ACPI_X86_CORE=y
CONFIG_ACPI_MEMORY_HOTPLUG=y
CONFIG_ACPI_CPU_HOTPLUG=y
CONFIG_APIC=y
CONFIG_IOAPIC=y
CONFIG_MEM_HOTPLUG=y
CONFIG_NVDIMM=y
CONFIG_ACPI_NVDIMM=y
CONFIG_PCIE_PORT=y
CONFIG_PXB=y
CONFIG_VTD=y
CONFIG_AMD_IOMMU=y
CONFIG_SYS_BUS_DEBUG=y
CONFIG_PCI_LITE=y
CONFIG_ACPI_HW_REDUCED=y
CONFIG_9P=y
CONFIG_MMIO_INTERFACE=y
CONFIG_ACPI_X86_CORE=y
CONFIG_VHOST_NET=y
CONFIG_X86_VIRT=y
CONFIG_KVM_CLOCK=y

If there are still some CONFIG_XXX unused in virt platform, please let me know, thanks!

yangzhon commented:
Another issue is:

If we add new softmmu file, x86_64_virt-softmmu.mak, we also need  add new target, TARGET_X86_64_VIRT, in some source code files, because those files only support TARGET_X86_64.

rbradford commented:
Hi @yangzhon.

Can you work on making your changes in "virt: hw: make files build configurable" more upstreamable. I think at the very least this should involve splitting the changes into more logical changes, commit messages explaining why you're doing that. You should also look around to see if some of the other CONFIG_ options make better sense than adding a new one or even maybe it's work using the boolean conditionals to "and" two config options together.

It is my firm belief that such missing build control of blocks are upstream gaps and so should be readily accepted upstream. As such I think this work should only focus on PC and Q35 directly from the 3.0 tag and then the topic/virt-x86 could be rebased on top of that.

yangzhon commented:
@sameo @rbradford @mcastelino

Please review, thanks!

yangzhon commented:
@sameo @rbradford @mcastelino

Add new x86_64_virt-softmmu build and TARGET_X86_64=y for all default TARGET_X86_64 build.
Tomorrow, i will remove some CONFIG_XXX in the default-configs/x86_64_virt-softmmu.mak for virt platform.

yangzhon commented:
@sameo @rbradford @mcastelino

The patchset made hw/ directory configurable and the pc/q35 platform code can be compiled out from virt binary. The minimal x86_64_virt-softmmu.mak is still on the way, and there are lots of work to do.

There are two unclear issues
(1). x86_64_virt-softmmu.mak will be minimal mak file, from current USB porting, i found we also need add --disable-usb during the configure command.   This will result in another issue, if the customer use the ./configure to do full build, which will generate errors in virt platform build because minimal virt mak file need co-work with some "--disable-xxx" commands.

(2) about the VGA
There are some kinds of VGA, which kind of VGA will be supported ? vfio-vga, pci-vga and virtio-vga?

rbradford commented:
@yangzhon virt platform will not support USB or VGA.

rbradford commented on default-configs/hw-common.mak:
```diff
@@ -0,0 +1 @@
+CONFIG_9P_FS=y
```
There is no need for this. 9pfs is already controlled by:

in hw/Makefile.objs:

```
devices-dirs-$(call land,$(CONFIG_VIRTFS),$(call lor,$(CONFIG_VIRTIO),$(CONFIG_XEN))) += 9pfs/
```

rbradford commented on default-configs/hw-common.mak:
```diff
@@ -1 +1,4 @@
 CONFIG_9P_FS=y
+CONFIG_BLOCK=y
```
block.c is needed for everyone so I don't think it ever makes sense to make this conditional.

rbradford commented on default-configs/hw-common.mak:
```diff
@@ -2,3 +2,4 @@ CONFIG_9P_FS=y
 CONFIG_BLOCK=y
 CONFIG_CDROM=y
 CONFIG_HARD_DISK=y
+CONFIG_UNIMP_DEV=y
```
I think this should be CONFIG_UNIMPLEMENTED

rbradford commented on default-configs/hw-common.mak:
```diff
@@ -3,3 +3,4 @@ CONFIG_BLOCK=y
 CONFIG_CDROM=y
 CONFIG_HARD_DISK=y
 CONFIG_UNIMP_DEV=y
+CONFIG_PCI_SHPC=y
```
I don't think this should be in hw-common.mak. Only in those files that use CONFIG_PCI=y

sameo replied:
Agreed. It should be in pci.mak imho.

rbradford commented on hw/pcmcia/Makefile.objs:
```diff
@@ -1,2 +1,2 @@
-common-obj-y += pcmcia.o
```
You should make this change at the hw/Makefile.objs level.

rbradford commented on default-configs/hw-common.mak:
```diff
@@ -4,3 +4,4 @@ CONFIG_CDROM=y
 CONFIG_HARD_DISK=y
 CONFIG_UNIMP_DEV=y
 CONFIG_PCI_SHPC=y
+CONFIG_PCMCIA=y
```
This shouldn't be here. pcmcia.c is a very basic object definition. You should put this in the config file where CONFIG_PXA2XX is turned on

rbradford commented on hw/bt/Makefile.objs:
```diff
@@ -1,3 +1,3 @@
-common-obj-y += core.o l2cap.o sdp.o hci.o hid.o
-common-obj-y += hci-csr.o
+common-obj-$(CONFIG_BT) += core.o l2cap.o sdp.o hci.o hid.o
```
I think CONFIG_BLUEZ from the configure script should be used instead.

rbradford commented on default-configs/hw-common.mak:
```diff
@@ -6,3 +6,5 @@ CONFIG_UNIMP_DEV=y
 CONFIG_PCMCIA=y
 CONFIG_BT=y
+CONFIG_HID_DEV=y
+CONFIG_PS2_DEV=y
```
This should only be on targets that only need it...I guess only for x86_64 and i386.

rbradford commented on default-configs/hw-common.mak:
```diff
@@ -6,3 +6,5 @@ CONFIG_UNIMP_DEV=y
 CONFIG_PCI_SHPC=y
 CONFIG_PCMCIA=y
 CONFIG_BT=y
+CONFIG_HID_DEV=y
```
Would be worth looking around what this virtual device is used for.

rbradford commented on hw/nvram/Makefile.objs:
```diff
@@ -1,7 +1,7 @@
 common-obj-$(CONFIG_DS1225Y) += ds1225y.o
-common-obj-y += eeprom93xx.o
+common-obj-$(CONFIG_93XX_EEPROM) += eeprom93xx.o
```
This is used by a few virtual devices. You should grep the source code and OR them rather than creating a new one.

rbradford commented on hw/nvram/Makefile.objs:
```diff
@@ -1,7 +1,7 @@
 common-obj-$(CONFIG_I2C) += eeprom_at24c.o
 common-obj-y += fw_cfg.o
-common-obj-y += chrp_nvram.o
+common-obj-$(CONFIG_NVRAM) += chrp_nvram.o
```
Ditto. This is used by just a few other virtual devices so you don't need to add a new config option.

rbradford commented on default-configs/hw-common.mak:
```diff
@@ -1 +1,4 @@
 CONFIG_9P_FS=y
+CONFIG_BLOCK=y
+CONFIG_CDROM=y
```
I think you will also need to add checks to vl.c to handle this not being present.

rbradford commented on default-configs/i386-softmmu.mak:
```diff
@@ -66,4 +66,5 @@ CONFIG_FW_CFG_DMA=y
 CONFIG_I2C=y
 CONFIG_SEV=$(CONFIG_KVM)
 CONFIG_VTD=y
+CONFIG_ACPI_X86_PIXX=y
```
I think adding a CONFIG_X86_PIIX would be better.

rbradford commented on hw/acpi/Makefile.objs:
```diff
@@ -1,5 +1,6 @@
 ifeq ($(CONFIG_ACPI),y)
-common-obj-$(CONFIG_ACPI_X86) += core.o piix4.o pcihp.o link.o
+common-obj-$(CONFIG_ACPI_X86) += core.o link.o
+common-obj-$(CONFIG_ACPI_X86_PIXX) += piix4.o pcihp.o
```
And then CONFIG_ACPI_X86 && CONFIG_X86_PIIX

rbradford commented on hw/audio/Makefile.objs:
```diff
@@ -15,4 +15,4 @@ common-obj-$(CONFIG_CS4231) += cs4231.o
 common-obj-$(CONFIG_MILKYMIST) += milkymist-ac97.o
 
-common-obj-y += soundhw.o
+common-obj-$(CONFIG_AUDIO) += soundhw.o
```
Just removing this will break vl.c.

rbradford commented on hw/cpu/Makefile.objs:
```diff
@@ -2,4 +2,4 @@ obj-$(CONFIG_ARM11MPCORE) += arm11mpcore.o
 obj-$(CONFIG_A9MPCORE) += a9mpcore.o
 obj-$(CONFIG_A15MPCORE) += a15mpcore.o
-common-obj-y += core.o
+common-obj-$(CONFIG_ARM_CORE) += core.o
```
There is nothing ARM specific about this. This file should always be included in the build and not made conditional

rbradford commented on hw/intc/Makefile.objs:
```diff
@@ -20,7 +20,7 @@ common-obj-$(CONFIG_ARM_GIC) += arm_gicv3_dist.o
 common-obj-$(CONFIG_ARM_GIC) += arm_gicv3_its_common.o
 common-obj-$(CONFIG_OPENPIC) += openpic.o
-common-obj-y += intc.o
+common-obj-$(CONFIG_INT_CTRL) += intc.o
```
This a trivial file. There is no harm in always including it.

rbradford commented on default-configs/hw-common.mak:
```diff
@@ -12,3 +12,4 @@ CONFIG_93XX_EEPROM=y
 CONFIG_NVRAM=y
 CONFIG_AUDIO=y
 CONFIG_INT_CTRL=y
+CONFIG_PCI_BRIDGE=y
```
What platforms actually need this PCI-PCI bridge?

rbradford commented on hw/pci-host/Makefile.objs:
```diff
@@ -1,4 +1,4 @@
-common-obj-y += pam.o
+common-obj-$(CONFIG_PAM) += pam.o
```
Use an OR of CONFIG_Q35 and CONFIG_PIIX (only files that use it)

rbradford commented on default-configs/hw-common.mak:
```diff
@@ -14,3 +14,4 @@ CONFIG_AUDIO=y
 CONFIG_INT_CTRL=y
 CONFIG_PCI_BRIDGE=y
 CONFIG_PAM=y
+CONFIG_SCSI=y
```
Already in many config files.

rbradford commented on hw/scsi/Makefile.objs:
```diff
@@ -1,5 +1,5 @@
-common-obj-y += scsi-disk.o
-common-obj-y += scsi-generic.o scsi-bus.o
+common-obj-$(CONFIG_SCSI) += scsi-disk.o
+common-obj-$(CONFIG_SCSI) += scsi-generic.o scsi-bus.o
```
Handled already in hw/Makefile.objs
```
devices-dirs-$(CONFIG_SCSI) += scsi/
```

yangzhon committed:
hw: make hw directory build configurable

Add one new hw-common.mak, which will add new
CONFIG_XXX for some built in files. Add this new
mak file to all softmmu make file, which will not
impact other softmmu build.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: make block build configurable

Add the new CONFIG_BLOCK{CDROM,HARD_DISK} to make
hw/block build configurable.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: make misc build configurable

Add new CONFIG_UNIMP_DEV for misc/unimp.c and make
it build configurable.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: make pcmcia build configurable

Add new CONFIG_PCMCIA for pcmcia and make it
build configurable.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: make bt build configurable

Add new CONFIG_BT for hw/bt directory and
make this directory build configurable.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: make HID/PS2 build configurable

Add new CONFIG_XXX for hid and ps2 devices and make
those files build configurable.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: make nvram build configurable

Add new CONFIG_XXX for nvram devices and make
those files build configurable.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: make acpi files build configurable

Split out the pixx related file with new CONFIG
in the hw/acpi.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: make audio file build configurable

Add new CONFIG_AUDIO for audio file in hw/audio and
make this file build configurable.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: make cpu file build configurable

The core.c file in hw/cpu is arm core file and
add the CONFIG in arm-softmmu.mak for arm.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: change CONFIG definition for ich9

Use the CONFIG_ACPI_X86_ICH to replace CONFIG_ACPI_X86
for more accurate definition.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: make i386 related files configurable

Add new CONFIGs for files in hw/i386 and change
related softmmu make file.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: make pci_bridge_dev.c build configurable

Add new CONFIG for pci_bridge_dev.c and make it
build configurable.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: make pam build configurable

pam is only used in pc/q35 and make it build
configurable.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: make usb build configurable

Make left usb files build configurable.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: make core build configurable

Add CONFIG for left core file in hw/core to make
this build configurable.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
virt: softmmu: make new x86_64_virt-softmmu.mak

Support new x86_64_virt-softmmu with x86_64_virt-softmmu.mak
and the new configure command as below:
./configure --target-list=x86_64_virt-softmmu

If we build x86_64_virt target softmmu binary, the TARGET_X86_64
is also enabled for some files build. like this, we do not need
change any code for TARGET_X86_64_VIRT.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: split pc.c file

The hw/i386/pc.c file is almost for pc/q35 platform and
split it to pc.c and x86.c. x86.c is common file for all
x86 platform. Change the CONFIG definition for pc.c and
compile out this file if we build virt platform.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: compile out q35/pc in virt platform

Remove the CONFIGs in the x86_64_virt-softmmu.mak and
compile out pc/q35 codes in virt platform.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: minimal x86_64_virt-softmmu.mak

This is minimal config file for less hw emulation code.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
virt: configure: x86_64_virt is only supported in target build

Since x86_64_virt target build use minimal CONFIGs and there are
some build error with full build. So this patch can make sure if
we do full build(./configure), the build do not include
qemu-system-x86_64_virt binary and qemu-system-x86_64 binary include
pc,q35 and virt build. If we only do virt target build
(./configure --target-list=x86_64_virt-softmmu),
qemu-system-x86_64_virt binary will be built.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
virt: vfio : add stubs function for display

Since virt platform do not support vga and vfio vga,
we add stubs function for vfio display in hw/vfio/pci.c.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
virt: hw: add stub irq functions for virt minimal build

Since vapic_report_tpr_access() and pic_read_irq() get
compiled out in virt platform, add stub functions for
those two functions build in other files.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
virt: bt: add stub functions for bt

Since hw/bt get compiled out, we need add stub functions
for those functions which are used in vl.c.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
virt: hw: add stub functions for vl.c

Since watchdog and audio hw get compiled out, we need add
stub functions for build in vl.c.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
virt: scsi: add stub functions for cdrom

Since cdrom get compiled out in minimal config and add
stub functions for cdrom which are used in scsi.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
virt: smbios : move fw_build_smbios() to smbios/smbios.c

fw_build_smbios() move to hw/smbios/smbios.c file for
CONFIG_SMBIOS and add stub function for disable smbios.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
virt: multiboot: add stub functions for multiboot

stub function needed once multiboot get compiled out.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
virt: hw : qmp rtc

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: make wdt build configurable

Add CONFIG for watchdog.c in hw/watchdog and make
it build configurable.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: make vga build configurable

Add CONFIGs for left vga files in hw/display and hw/vfio
to make those build configurable.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: make shpc.c build configurable

Add the new CONFIG_PCI_SHPC for pci/shpc.c and make
it build configurable.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: move build_madt to aml-build.c

Move build_madt() to aml-build.c file, that can make
hw/i386/acpi-build.c is only used for pc/q35.

Change build_madt() to acpi_build_madt() is only for
fixing build conflict with arm.

Signed-off-by: Yang Zhong <yang.zhong@intel.com>

sameo commented on default-configs/x86_64_virt-softmmu.mak:
```diff
@@ -0,0 +1,80 @@
+CONFIG_X86_Q35=y
+CONFIG_X86_VIRT=y
+CONFIG_KVM_VAPIC=y
+CONFIG_X86_PC_Q35_ACPI_BUILD=y
```
This one should be shorter, something along those lines:

```
# Default configuration for x86_64_virt-softmmu

# PCI config
CONFIG_PCI=y
CONFIG_PCI_BRIDGE=y
CONFIG_SERIAL=y
CONFIG_SERIAL_PCI=y
CONFIG_NVME_PCI=y

# virtio and vhost
CONFIG_VIRTIO=y
CONFIG_VIRTIO_9P=y
CONFIG_VIRTIO_BALLOON=y
CONFIG_VIRTIO_BLK=y
CONFIG_VIRTIO_CRYPTO=y
CONFIG_VIRTIO_GPU=y
CONFIG_VIRTIO_INPUT=y
CONFIG_VIRTIO_NET=y
CONFIG_VIRTIO_RNG=y
CONFIG_VIRTIO_SCSI=y
CONFIG_VIRTIO_SERIAL=y
CONFIG_VIRTIO_PCI=y
CONFIG_VHOST_NET=y
CONFIG_VHOST_SCSI=y
CONFIG_VHOST_USER_SCSI=$(call land,$(CONFIG_VHOST_USER),$(CONFIG_LINUX))
CONFIG_VHOST_USER_BLK=$(call land,$(CONFIG_VHOST_USER),$(CONFIG_LINUX))

# HW common
CONFIG_9P=y
CONFIG_9P_FS=y
CONFIG_BLOCK=y
CONFIG_CDROM=y
CONFIG_HARD_DISK=y
CONFIG_NVRAM=y
CONFIG_INT_CTRL=y
CONFIG_SCSI=y
CONFIG_TPM=y
CONFIG_TPM_UTIL=y

# Others
CONFIG_ACPI=y
CONFIG_ACPI_X86=y
CONFIG_ACPI_MEMORY_HOTPLUG=y
CONFIG_ACPI_CPU_HOTPLUG=y
CONFIG_ACPI_NVDIMM=y
CONFIG_APIC=y
CONFIG_IOAPIC=y
CONFIG_MEM_HOTPLUG=y
CONFIG_NVDIMM=y
CONFIG_PCIE_PORT=y
CONFIG_PXB=y
CONFIG_VTD=y
CONFIG_AMD_IOMMU=y
CONFIG_SYS_BUS_DEBUG=y
CONFIG_PCI_LITE=y
CONFIG_ACPI_HW_REDUCED=y
CONFIG_X86_VIRT=y
CONFIG_KVM_CLOCK=y
CONFIG_KVM_VAPIC=y
```

yangzhon commented:
@sameo @rbradford @mcastelino

I submitted one NEW PR #80 to replace this older one, drop this one, thanks!
